src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -644,6 +644,82 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
             }
         }
 
+        protected override void ApplyProjectChanges(ProjectChanges projectChanges)
+        {
+            base.ApplyProjectChanges(projectChanges);
+
+            var changedMappedDocuments = projectChanges.GetChangedDocuments().Where(id => IsMappedDocument(id));
+            if (changedMappedDocuments.Any())
+            {
+                ApplyChangesForMappedDocuments(projectChanges, changedMappedDocuments);
+            }
+
+            bool IsMappedDocument(DocumentId documentId)
+            {
+                return this.CurrentSolution.GetDocument(documentId)?.Services?.GetService<ISpanMappingService>() != null;
+            }
+        }
+
+        private void ApplyChangesForMappedDocuments(ProjectChanges projectChanges, IEnumerable<DocumentId> changedMappedDocuments)
+        {
+            // Map all the original text changes to new text changes in the mapped files.
+            var _ = ArrayBuilder<(string fileName, TextChange textChange)>.GetInstance(out var mappedChangesBuilder);
+            foreach (var changedDocumentId in changedMappedDocuments)
+            {
+                var oldDoc = projectChanges.OldProject.GetDocument(changedDocumentId)!;
+                var newDoc = projectChanges.NewProject.GetDocument(changedDocumentId)!;
+
+                GetMappedTextChanges(newDoc, oldDoc, mappedChangesBuilder);
+            }
+
+            var mappedChanges = mappedChangesBuilder.ToImmutable();
+            if (mappedChanges.Any())
+            {
+                // Group the mapped text changes by file, then apply all mapped text changes to each file.
+                foreach (var changesForFile in mappedChanges.GroupBy(change => change.fileName))
+                {
+                    using var invisibleEditor = new InvisibleEditor(ServiceProvider.GlobalProvider, changesForFile.Key, GetHierarchy(projectChanges.ProjectId), needsSave: true, needsUndoDisabled: false);
+                    TextEditApplication.UpdateText(changesForFile.Select(change => change.textChange), invisibleEditor.TextBuffer, EditOptions.None);
+                }
+            }
+
+            void GetMappedTextChanges(CodeAnalysis.Document newDocument, CodeAnalysis.Document oldDocument, ArrayBuilder<(string, TextChange)> builder)
+            {
+                var mappingService = newDocument.Services.GetService<ISpanMappingService>();
+                (var mappedSpanResults, var textChanges) = _threadingContext.JoinableTaskFactory.Run<(ImmutableArray<MappedSpanResult>, ImmutableArray<TextChange>)>(async () =>
+                {
+                    var textChanges = (await newDocument.GetTextChangesAsync(oldDocument, CancellationToken.None).ConfigureAwait(false)).ToImmutableArray();
```

Comment:
what does this mean to be non-cancellable.  if we're doing IO (including network) this seems like this can effectively hang, right?

Reply:
It's possible we do file IO to get the text changes, but that is already the case for all apply changes, see http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Workspace.cs,1565
Mapping the spans could potentially do IO as well to read document contents, but none of these involve network calls

Unfortunately applying changes doesn't have a cancellation token (I believe because cancelling could leave us with half applied changes)